File: src/metadata.ts
import * as fs from 'fs/promises';
import { constants as fsConstants } from 'fs';
import ExifReader from 'exifreader';
import ffmpeg from 'fluent-ffmpeg';
import pLimit from 'p-limit';
import { parseExifDate } from './dateFormatter';
import { CONCURRENCY_LIMIT } from './constants';
import { MediaFile, MediaMetadata, DateSource, ProgressCallback } from './types';

/**
 * Checks if a file exists
 */
async function fileExists(filePath: string): Promise<boolean> {
  try {
    await fs.access(filePath, fsConstants.F_OK);
    return true;
  } catch {
    return false;
  }
}

/**
 * Extracts date from image EXIF metadata
 */
async function extractImageDate(filePath: string): Promise<{ date: Date; source: DateSource }> {
  try {
    const buffer = await fs.readFile(filePath);
    const tags = ExifReader.load(buffer, { expanded: true });
    
    // Access exif tags with type safety
    const exifTags = tags.exif as Record<string, { description?: string } | undefined> | undefined;
    
    // Try DateTimeOriginal first (when photo was actually taken)
    const dateTimeOriginal = exifTags?.DateTimeOriginal?.description;
    if (dateTimeOriginal) {
      const date = parseExifDate(dateTimeOriginal);
      if (date) return { date, source: 'exif' };
    }
    
    // Try CreateDate as fallback (also known as DateTimeDigitized in some cameras)
    const createDate = exifTags?.CreateDate?.description || exifTags?.DateTimeDigitized?.description;
    if (createDate) {
      const date = parseExifDate(createDate);
      if (date) return { date, source: 'exif' };
    }
    
    // Try DateTime as another fallback
    const dateTime = exifTags?.DateTime?.description;
    if (dateTime) {
      const date = parseExifDate(dateTime);
      if (date) return { date, source: 'exif' };
    }
  } catch {
    // EXIF extraction failed, fall back to file system dates
  }
  
  // Fallback to file system dates
  return extractFilesystemDate(filePath);
}

/**
 * Extracts date from video metadata using ffprobe
 */
async function extractVideoDate(filePath: string): Promise<{ date: Date; source: DateSource }> {
  return new Promise((resolve) => {
    ffmpeg.ffprobe(filePath, async (err, metadata) => {
      if (!err && metadata) {
        // Try to get creation_time from format tags
        const creationTime = metadata.format?.tags?.creation_time;
        if (creationTime) {
          const date = new Date(creationTime);
          if (!isNaN(date.getTime())) {
            resolve({ date, source: 'video_metadata' });
            return;
          }
        }
        
        // Try QuickTime specific tag (for MOV files from iPhones)
        const quicktimeDate = metadata.format?.tags?.['com.apple.quicktime.creationdate'];
        if (quicktimeDate) {
          const date = new Date(quicktimeDate);
          if (!isNaN(date.getTime())) {
            resolve({ date, source: 'video_metadata' });
            return;
          }
        }
      }
      
      // Fallback to file system dates
      const result = await extractFilesystemDate(filePath);
      resolve(result);
    });
  });
}

/**
 * Extracts date from file system metadata
 */
async function extractFilesystemDate(filePath: string): Promise<{ date: Date; source: DateSource }> {
  const stats = await fs.stat(filePath);
  
  // Prefer birthtime (file creation date) if available
  if (stats.birthtime && stats.birthtime.getTime() > 0) {
    return { date: stats.birthtime, source: 'file_birthtime' };
  }
  
  // Fall back to modification time
  return { date: stats.mtime, source: 'file_mtime' };
}

/**
 * Extracts metadata from a single media file
 */
export async function extractMetadata(file: MediaFile): Promise<MediaMetadata> {
  const result = file.type === 'image' 
    ? await extractImageDate(file.path)
    : await extractVideoDate(file.path);
  
  return {
    file,
    createdDate: result.date,
    dateSource: result.source
  };
}

/**
 * Extracts metadata from multiple media files with parallel processing
 */
export async function extractAllMetadata(
  files: MediaFile[],
  onProgress?: ProgressCallback
): Promise<MediaMetadata[]> {
  const limit = pLimit(CONCURRENCY_LIMIT);
  let completed = 0;
  
  const promises = files.map(file => 
    limit(async () => {
      const metadata = await extractMetadata(file);
      completed++;
      onProgress?.(completed, files.length);
      return metadata;
    })
  );
  
  return Promise.all(promises);
}

File: src/metadata.rs
//! Parallel metadata extraction over all discovered media files.
//!
//! Mirrors the date-source priority of the legacy TypeScript pipeline:
//!   * images   : EXIF (`DateTimeOriginal -> CreateDate -> DateTime`) → fs
//!   * mp4/mov  : QuickTime keys → mvhd → fs
//!   * other vid: ffprobe (creation_time, qt key) → fs
//!
//! Replaces `src/metadata.ts`'s `pLimit(10)` cap with a real rayon thread pool
//! sized to the available CPU count.
//!
//! Files for which no real capture date *and* no usable filesystem timestamp
//! can be obtained are returned separately as `UnsortableMedia` so the caller
//! can route them into the `unsortable/` folder and surface a per-file reason
//! in the summary.

use crate::exif_image::extract_image_date;
use crate::extensions::{uses_ffprobe, uses_mp4parse};
use crate::ffprobe_fallback::extract_via_ffprobe;
use crate::mp4_meta::extract_mp4_date;
use crate::types::{DateSource, MediaFile, MediaMeta, MediaType, UnsortableMedia};
use chrono::{DateTime, Local, TimeZone};
use rayon::prelude::*;
use std::path::Path;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, Mutex};

pub trait MetadataProgress: Send + Sync {
    fn on_completed(&self, _meta: &MediaMeta) {}
    fn on_unsortable(&self, _file: &UnsortableMedia) {}
}

#[allow(dead_code)]
pub struct NoopProgress;
impl MetadataProgress for NoopProgress {}

#[derive(Debug, Default)]
pub struct ExtractResult {
    pub metas: Vec<MediaMeta>,
    pub unsortable: Vec<UnsortableMedia>,
}

pub fn extract_all(
    files: Vec<MediaFile>,
    progress: Arc<dyn MetadataProgress>,
) -> ExtractResult {
    let done = AtomicU64::new(0);
    let unsortable: Mutex<Vec<UnsortableMedia>> = Mutex::new(Vec::new());

    let metas: Vec<MediaMeta> = files
        .into_par_iter()
        .filter_map(|file| {
            let outcome = extract_one(file);
            done.fetch_add(1, Ordering::Relaxed);
            match outcome {
                Ok(meta) => {
                    progress.on_completed(&meta);
                    Some(meta)
                }
                Err(bad) => {
                    progress.on_unsortable(&bad);
                    unsortable.lock().expect("unsortable mutex").push(bad);
                    None
                }
            }
        })
        .collect();

    ExtractResult {
        metas,
        unsortable: unsortable.into_inner().expect("unsortable mutex"),
    }
}

fn extract_one(file: MediaFile) -> Result<MediaMeta, UnsortableMedia> {
    let metadata_attempt = match file.media_type {
        MediaType::Image => extract_image_date(&file.path),
        MediaType::Video => {
            if uses_mp4parse(&file.extension) {
                extract_mp4_date(&file.path)
            } else if uses_ffprobe(&file.extension) {
                extract_via_ffprobe(&file.path)
            } else {
                None
            }
        }
    };

    if let Some((created_date, date_source)) = metadata_attempt {
        return Ok(MediaMeta {
            file,
            created_date,
            date_source,
        });
    }

    if let Some((created_date, date_source)) = filesystem_date(&file.path) {
        return Ok(MediaMeta {
            file,
            created_date,
            date_source,
        });
    }

    Err(UnsortableMedia {
        reason: unsortable_reason(&file),
        file,
    })
}

fn unsortable_reason(file: &MediaFile) -> String {
    let kind = match file.media_type {
        MediaType::Image => "no EXIF date in image",
        MediaType::Video if uses_mp4parse(&file.extension) => {
            "no creation date in MP4/MOV/M4V/3GP container"
        }
        MediaType::Video if uses_ffprobe(&file.extension) => {
            "ffprobe could not read a creation date (or ffprobe is not installed)"
        }
        MediaType::Video => "no metadata extractor available for this video format",
    };
    format!(
        "{kind}; filesystem birth/modify timestamps are also unreadable or zero"
    )
}

/// Falls back to filesystem timestamps in the same priority as the legacy code:
/// `birthtime` first if available and non-zero, then `mtime`. Returns `None`
/// only when both are unavailable or zero — in which case the file is treated
/// as unsortable rather than being silently bucketed into 1970-01-01.
fn filesystem_date(path: &Path) -> Option<(DateTime<Local>, DateSource)> {
    let meta = std::fs::metadata(path).ok()?;
    if let Ok(birth) = meta.created() {
        if let Some(dt) = system_to_local(birth) {
            if dt.timestamp() > 0 {
                return Some((dt, DateSource::FileBirthtime));
            }
        }
    }
    if let Ok(modified) = meta.modified() {
        if let Some(dt) = system_to_local(modified) {
            if dt.timestamp() > 0 {
                return Some((dt, DateSource::FileMtime));
            }
        }
    }
    None
}

fn system_to_local(t: std::time::SystemTime) -> Option<DateTime<Local>> {
    let dur = t.duration_since(std::time::UNIX_EPOCH).ok()?;
    let secs = dur.as_secs() as i64;
    let nanos = dur.subsec_nanos();
    Local.timestamp_opt(secs, nanos).single()
}

// Kept for documentation: the legacy behaviour synthesised an epoch
// timestamp when nothing else worked. We intentionally no longer do that;
// see `filesystem_date` returning `Option`.
#[allow(dead_code)]
fn _legacy_epoch_fallback() -> DateTime<Local> {
    Local
        .timestamp_opt(0, 0)
        .single()
        .unwrap_or_else(|| Local.with_ymd_and_hms(1970, 1, 1, 0, 0, 0).single().unwrap())
}
